packages/explat-client/src/internal/state.ts

```diff
@@ -3,20 +3,30 @@
  */
 import type { ExperimentAssignment } from '../types';
 import * as Validations from './validations';
-import * as ExperimentAssignments from './experiment-assignments';
 
-const experimentNameToExperimentAssignment: Record< string, ExperimentAssignment | undefined > = {};
+type Store = {
+	experimentNameToExperimentAssignment: Record< string, ExperimentAssignment | undefined >;
+};
 
+export function createStore(): Store {
+	return {
+		experimentNameToExperimentAssignment: {},
+	};
+}
```

Comment:
Can please explain the need for this more? If it's meant to be passed in by clients, shouldn't it be a part of the config?

Reply:
Ah, I think I see what's going on given #49832. Is it so that each client would have its own store, but the store is still hidden from the client?

If we're going with this, I think it'd be cleaner for this to be more of a class-like interface, as the `store` is mutable. That is, create a store object with with methods `storeExperimentAssignment()` and `retrieveExperimentAssignment()`.

Reply:
It is used like this: 
https://github.com/Automattic/wp-calypso/pull/48475/files#diff-6ccc8c485f7dc1c524767b67ae2efa00edf70f04ab35c903a9f0287b644d507aR63

Not meant to be passed in by clients.

Reply:
Oops I didn't see your last comment: Yes we could be using a class here (and also for the ExPlat client), but I think this is simplest, will discourage adding extra state, and makes unit testing a breeze.

Reply:
I disagree about it being simplest, especially when looking at the way it's used. I really don't understand the aversion to using classes when they make sense...

`store.storeExperimentAssignment( experimentAssignment )` is clearer to me than `Store.storeExperimentAssignment( store, experimentAssignment )`, as it's more self-documenting that it operates on the `store`, which is a mutable object. It's also in line with browser APIs like `localStorage`.

Reply:
> I really don't understand the aversion to using classes when they make sense...

The way I see it is that classes are inherently more complex than functions - they add another layer of modelling and most importantly developers are more likely to misuse them (adding state, adding methods, mixing up purposes because it is easier to get at the state from within, needing to add more methods to utilise private methods). 

If we aren't using features that classes provide over functions like inheritance, or run-time polymorphism and we aren't using it often and everywhere, to me the added "convenience" of using them doesn't outweigh the added complexity.

So as a general rule, where they are providing the same features except for less argument to functions, I like to avoid classes and prefer functions.

> store.storeExperimentAssignment( experimentAssignment ) is clearer to me than Store.storeExperimentAssignment( store, experimentAssignment ), as it's more self-documenting that it operates on the store, which is a mutable object.

It is true we are passing a mutable state reference in, but I see this as a pretty small difference, I don't think any Javascript dev is going to trip over it.

> It's also in line with browser APIs like localStorage.

Browser APIs like localStorage are often the result of years of legacy, compromise and committees. We should be working from modern principles and practices in Javascript. 

It is also a pretty simple tool that isn't to be used externally so familiarity shouldn't be a problem.

Reply:
I'm sorry, but I don't buy it. What you're framing as "modern principles and practices in JavaScript" in this case looks a lot like abstract data types I saw implemented in C in my computer science degree, before they introduced us to C++. And I won't even get into the JS world's tendency to reinvent the wheel every month with a shiny new framework. :trollface: 

But anyway, focusing on the matter at hand, it's worth looking at what the implementation would look like:

```typescript
class Store {
  private experimentNameToExperimentAssignment: Record< string, ExperimentAssignment | undefined > = {}

  public storeExperimentAssignment( experimentAssignment: ExperimentAssignment ): void {
	  Validations.validateExperimentAssignment( experimentAssignment );
    const previousExperimentAssignment =
      this.experimentNameToExperimentAssignment[ experimentAssignment.experimentName ];
    if (
      previousExperimentAssignment &&
      experimentAssignment.retrievedTimestamp < previousExperimentAssignment.retrievedTimestamp
    ) {
      throw new Error(
        'Trying to store an older experiment assignment than is present in the store, likely a race condition.'
      );
    }

    this.experimentNameToExperimentAssignment[ experimentAssignment.experimentName ] = experimentAssignment;
  }

  public retrieveExperimentAssignment( experimentName: string ): ExperimentAssignment | undefined {
    return this.experimentNameToExperimentAssignment[ experimentName ];
  }
}
```

This is more than "less arguments to functions". It:

* Self-documents the mutability of the underlying object.
* Obviates the need for a custom function like `createStore()` &ndash; we get the constructor for "free" from TypeScript.
* Clearly links the store and retrieve method to each other and to the Store object they operate on.

I don't see why we'd avoid using this implementation here, especially as developer misuse is unlikely to be an issue. It seems like a perfect use case for a simple class to me.

Reply:
> I'm sorry, but I don't buy it. What you're framing as "modern principles and practices in JavaScript" in this case looks a lot like abstract data types I saw implemented in C in my computer science degree, before they introduced us to C++. And I won't even get into the JS world's tendency to reinvent the wheel every month with a shiny new framework. :trollface:

This was in reply to your suggestion of basing it off of the `localStorage` browser API, rather I believe I am providing you with a first principles based argument and I think we should stick to that.

> I don't see why we'd avoid using this implementation here, especially as developer misuse is unlikely to be an issue. It seems like a perfect use case for a simple class to me.

It seems to me that all of the points you list (except for "getting the constructor for free" which I think doesn't have much material benefit) are in-fact for the purpose of making it harder for developers to misuse the code: self-documenting the mutability, linking the functions together and to the data. So to me it looks like this discussion does come down to whether classes or functions prevent developer misuse more.

I am happy to go with your call but I still believe that using classes in this kind of case creates more issues than they solve - including the benefits you mention.

Reply:
> This was in reply to your suggestion of basing it off of the `localStorage` browser API, rather I believe I am providing you with a first principles based argument and I think we should stick to that.

Yeah, I disagree that the "first principles" you cite are indeed universally accepted, which means that they're not first principles (happy to be proven wrong, though). I think that the `localStorage` browser API (or more generally, [the `Storage` API it implements](https://developer.mozilla.org/en-US/docs/Web/API/Storage)) would be implemented pretty similarly these days, as it's a fairly natural way of modelling it in languages that support classes. As noted, having a collection of functions operating on an opaque data structure reminds me more of languages that don't support classes, like C.

> So to me it looks like this discussion does come down to whether classes or functions prevent developer misuse more.

I don't see it as misuse &ndash; it's more a matter of clarity.

> I am happy to go with your call but I still believe that using classes in this kind of case creates more issues than they solve - including the benefits you mention.

Yeah, it's a matter of opinion to an extent, but it feels a bit like just a general aversion to using classes based on the potential future changes to the code, which I don't think is supported by much concrete evidence. Future misuse and messiness are possible with any approach, so it's more about the current readability of the code to me.

Let's get another opinion from @withinboredom on the readability of either solution and go with that.

Reply:
Holy discussion Batman!

> I really don't understand the aversion to using classes when they make sense...

There are [no such things as classes in older Javascript](https://www.typescriptlang.org/play?target=1#code/MYGwhgzhAEDKAuB7ATgU2gbwFDWgB2QEsA3MedVADz1SIFtUA7eAOTAYBVEBRa2whswCCUQgHNGg+AC5oAJVTAUAEwA80CPCKMxAGmhhGAT2gAfaAFdGy1ADNCjVMugA+aAF5MAXyw58FgCMQQmANJDReGnomeBEIcUkYgApoKiiBGLiEqVlDEwBKWWJEQmdsAEhcAHoqgHcHZURagDpSYOUyVEj+KSyJKRS0nszRfpjofIBuP1wlRk18NGJCRAsIbujhUcTmDxncaHgAC0IIZqHN1nZULg2Mrfix5gBtVL5Lvp34c-f7q4ZoABdaYHaCEWzQJL7XAEVDLVbrX69bZSaAAMjR0Le6WRjy+zTQWkIcKcHAymnYeGg6lh8LWd1x2RiBNQRJJyjJDApdDw+3ymCxx2QTWgjlq0G4yGFyChoNBAHIOMgjA4xIdEGEUOhDNBECAbMhscNdpA8ajjjrTotUBBxg5Dkd0JotfpggBrVAgExgaDIMDAdBzZSEeArRjNeVYqb7Hz7Y6nH44mJsTg8JEjM0xV4XP6fKSJ43-dCAjxGj4omIg6CxmGBYKhQlEEkMjNM5iDdPMFOoWSabRqwoGYxmSzWOwOJwC0GEizIRgOhM5qTd26d2IVl5lv7doFVnxeIA) and in newer browsers that support the `class` keyword, [it's really just syntactical sugar](https://medium.com/@naveenkarippai/debunking-fake-classes-in-javascript-78f67a6b5c96#:~:text=Dec%2012%2C%202016%C2%B77%20min%20read&text=TL%3BDR%3A%20ECMAScript6%20introduced%20class,compared%20to%20complex%20Function%20Constructors.). Javascript is a prototypical language and doesn't have inheritance like in OOP based languages. The `class` keyword in TypeScript allows one to code as though it does by transpiling any specific logic needed for it.

Using classes may add significant complexity to the code that is actually executed by the browser and slow things down. They should be used sparingly unless the use-case calls for it.

> Let's get another opinion from @withinboredom on the readability of either solution and go with that.

FWIW, I think we should go with a non-class solution for storing the state. If a developer desires, they can store the object in `localstorage` if they want to, or send it over the wire somewhere else. If it's a class with behavior, it makes things more complicated when loading it from `localstorage`, `IndexedDb`, or wherever.

This all looks like idiomatic Javascript to me.

Reply:
> Yeah, I disagree that the "first principles" you cite are indeed universally accepted, which means that they're not first principles (happy to be proven wrong, though).

I mean first principles as an approach to arguments based on deductions and assumptions, it isn't necessarily about agreeing on assumptions. The main body of [my reply here](https://github.com/Automattic/wp-calypso/pull/49829#discussion_r572616530) rather than the rebuttal of your point. 

> I think that the localStorage browser API (or more generally, the Storage API it implements) would be implemented pretty similarly these days, as it's a fairly natural way of modelling it in languages that support classes. As noted, having a collection of functions operating on an opaque data structure reminds me more of languages that don't support classes, like C.

If it is done there it doesn't necessarily make it right here, and just because using data structures in C isn't great that doesn't mean it applies here. (And lots of people still use C like that, particularly in the open source world.)

> I don't see it as misuse – it's more a matter of clarity.

For what purpose is clarity useful if not for aiding future development?

> it feels a bit like just a general aversion to using classes based on the potential future changes to the code, which I don't think is supported by much concrete evidence.

I feel like I have given pretty strong reasons not to use classes in these trivial cases - complexity from modelling and problems of developer misuse outweighing the benefits of self-documentation and clarity. Given that there is no concrete evidence on either side I don't see anything wrong with preferring functions over classes for trivial cases.

> Let's get another opinion from @withinboredom on the readability of either solution and go with that.

Why is readability the target and not maintainability?

Reply:
> If it is done there it doesn't necessarily make it right here, and just because using data structures in C isn't great that doesn't mean it applies here. (And lots of people still use C like that, particularly in the open source world.)

That's missing my point. C's syntax is limited. In languages that have richer syntax, its worth using it to simplify things.

> Why is readability the target and not maintainability?

Readability and maintainability are pretty synonymous.

> For what purpose is clarity useful if not for aiding future development?

Yes, clarity is useful for aiding future development. I just don't buy the dangers of "misuse", which are avoided by owning and reviewing code.

> I feel like I have given pretty strong reasons not to use classes in these trivial cases - complexity from modelling and problems of developer misuse outweighing the benefits of self-documentation and clarity. Given that there is no concrete evidence on either side I don't see anything wrong with preferring functions over classes for trivial cases.

Well, we can agree to disagree on this one, and go with @withinboredom's opinion as the tie breaker. :slightly_smiling_face: